### PR TITLE
umm, i can explain

### PR DESCRIPTION
**Pull request by time-killer-games**
tl;dr. In the process of me trying to fix dialogs, I accidentally broke them in a different way, while not actually fixing anything. 

https://github.com/enigma-dev/enigma-dev/commit/a019ab6c06cd922d264abef7f0f3a154d66c867d

When ci passes please merge. Posix threads are a more technically correct approach than using a fork anyway.

On Unix likes, regardless of the default shell executable that's actually being used, getting the first argument in the command line will return "/bin/sh" and this has been confirmed from testing. The problem was I wasnt getting a pid match with the launched process and kdialog/zenity because it was giving me the process id of the default shell instead; this code filters out the default shell to ensure a pid match so we can verify the topmost window is owned by the correct process.

### CODE REVIEW

**Comment by time-killer-games** [comment]
@JoshDreamland aight stinky poo brain, mudge pls

**Comment by time-killer-games** [comment]
@JoshDreamland hey boi

**Comment by time-killer-games** [comment]
@fundies @JoshDreamland umm don't cancel my ci kthx

**Comment by fundies** [comment]
> @fundies @JoshDreamland umm don't cancel my ci kthx

I was doing things and didn't want to wait. You can always close/reopen to trigger it again

**Comment by time-killer-games** [comment]
It's ok, I was just asking if it could wait this time if the same were to occur this time around.

File: ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp [hunk 1]
```diff
@@ -146,14 +161,86 @@ static pid_t process_execute(const char *command, int *infp, int *outfp) {
...
+  int cntp; if (kinfo_proc *proc_info = kinfo_getallproc(&cntp)) {
+    for (int j = 0; j < cntp; j++) {
+      if (proc_info[j].ki_ppid == parentProcId) {
+        vec.push_back(proc_info[j].ki_pid); i++;
+      }
+    }
+    free(proc_info);
+  }
+  #endif
+  *procId = (pid_t *)malloc(sizeof(pid_t) * vec.size());
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp:206]: Why are you copying the vector to a new buffer rather than just returning the vector?

**Review by JoshDreamland — COMMENTED** [review]
We can't keep adding dependency after dependency to accommodate something like displaying text boxes and prompting for strings. We'll go ahead and add this one, but if we get a bunch of user complaints about the new dep, I'm going to revert the default widget system back to None.

File: ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp [hunk 2]
```diff
@@ -146,14 +161,86 @@ static pid_t process_execute(const char *command, int *infp, int *outfp) {
...
-  while (pid_from_wid(display, wid) != pid) {
-    wid = wid_from_top(display);
+  Display *display = XOpenDisplay(nullptr);
+  Window wid = WidFromTop(display);
+  pid_t child = PidFromPpidRecursive((pid_t)(std::intptr_t)pid);
+  while (true) {
+    wid = WidFromTop(display);
+    if (PidFromWid(display, wid) == child) {
+      break;
+    }
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp:234]: This looks like a busy wait; I recommend adding a sleep or yield to avoid tying up the CPU while the window launches.

**Comment by time-killer-games** [comment]
> We can't keep adding dependency after dependency to accommodate something like displaying text boxes and prompting for strings. We'll go ahead and add this one, but if we get a bunch of user complaints about the new dep, I'm going to revert the default widget system back to None.

I don't see the pr changing the default widget system to xlib. If my previous or did that, it was a mistake on my end. I meant to keep it sent to None, and I wouldn't have insisted on you merging it in that state if I knew it did that, because I thought we already had it accomplished setting it to None was the proper thing to do.

File: ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp (hunk 2 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp:227]: Is five milliseconds good?

File: ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp (hunk 1 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/xlib/dialogs.cpp]: Sorry, this is based on code where i was already doing that and I forgot to make it C++ abi and not C.

**Review by JoshDreamland — APPROVED** [review]
Probably just got committed by accident at some point along the way. Without forcing this onto users as a default, I have no problems including it as a dep. Users actually expecting to launch dialogs this way will already need to install Zenity/KDialog, so this extra library isn't a big deal.
